hoomd/mpcd/CosineChannelGeometry.h

```diff
@@ -0,0 +1,317 @@
+             * the wall, the current position inside of the  wall, and the new position exactly at
+             * the wall are on a straight line.
+             */
+            z0 = -(pos.x - dt * vel.x - x0) * vel.z / vel.x + (pos.z - dt * vel.z);
+
+            // Newton's method sometimes failes to converge (close to saddle points, df'==0, bad
+            // initial guess,overshoot,..) catch all of them here and do bisection if Newthon's
+            // method didn't work
+            Scalar lower_x = fmin(pos.x - dt * vel.x, pos.x);
+            Scalar upper_x = fmax(pos.x - dt * vel.x, pos.x);
+
+            // found intersection is NOT in between old and new point, ie intersection is
+            // wrong/inaccurate. do bisection to find intersection - slower but more robust than
+            // Newton's method
+            if (x0 < lower_x || x0 > upper_x)
+                {
+                counter = 0;
+                Scalar3 point1 = pos;            // final position at t+dt, outside of channel
+                Scalar3 point2 = pos - dt * vel; // initial position, inside of channel
+                Scalar3 point3 = 0.5 * (point1 + point2); // halfway point
+                Scalar fpoint3 = (m_amplitude * fast::cos(point3.x * m_wavenumber) + sign * m_H)
+                                 - point3.y; // value at halfway point, f(x)
+                // Note: technically, the presicion of Newton's method and bisection is slightly
+                // different, with bisection being less precise and slower convergence.
+                while (fabs(fpoint3) > target_precision && counter < max_iteration)
+                    {
+                    fpoint3 = (m_amplitude * fast::cos(point3.x * m_wavenumber) + sign * m_H)
+                              - point3.y;
+                    // because we know that point1 outside of the channel and point2 is inside of
+                    // the channel, we only need to check the halfway point3 - if it is inside,
+                    // replace point2, if it is outside, replace point1
+                    if (isOutside(point3) == false)
+                        {
+                        point2 = point3;
+                        }
+                    else
+                        {
+                        point1 = point3;
+                        }
+                    point3 = 0.5 * (point1 + point2);
+                    ++counter;
+                    }
+                // final point3 == intersection
+                x0 = point3.x;
+                y0 = (m_amplitude * fast::cos(x0 * m_wavenumber) + sign * m_H);
+                z0 = -(pos.x - dt * vel.x - x0) * vel.z / vel.x + (pos.z - dt * vel.z);
+                }
+            }
+
+        // Remaining integration time dt is amount of time spent traveling distance out of bounds.
+        Scalar3 pos_new = make_scalar3(x0, y0, z0);
+        dt = fast::sqrt(dot((pos - pos_new), (pos - pos_new)) / dot(vel, vel));
```

Comment:
I would use a slow sqrt here to make sure this remaining time is accurate.
```suggestion
        dt = slow::sqrt(dot((pos - pos_new), (pos - pos_new)) / dot(vel, vel));
```